### PR TITLE
Fix racetest failures in envoy/v2 TestEds [release-1.0]

### LINKED ISSUE #11621: Racetest failure: istio.io/istio/pilot/pkg/proxy/envoy/v2 TestEds
[https://gubernator.k8s.io/build/istio-circleci/presubmit/racetest/320345](https://gubernator.k8s.io/build/istio-circleci/presubmit/racetest/320345)

```
==================
WARNING: DATA RACE
Read at 0x00c42061e448 by goroutine 104:
  istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).periodicRefreshMetrics()
      /go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/discovery.go:234 +0x15e

Previous write at 0x00c42061e448 by goroutine 292:
  [failed to restore the stack]

Goroutine 104 (running) created at:
  istio.io/istio/pilot/pkg/proxy/envoy/v2.NewDiscoveryServer()
      /go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/discovery.go:149 +0x27a
  istio.io/istio/pilot/pkg/bootstrap.(*Server).initDiscoveryService()
      /go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:779 +0x640
  istio.io/istio/pilot/pkg/bootstrap.NewServer()
      /go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:227 +0x26f
  istio.io/istio/tests/util.setup()
      /go/src/istio.io/istio/tests/util/pilot_server.go:164 +0x3dd
  istio.io/istio/tests/util.EnsureTestServer()
      /go/src/istio.io/istio/tests/util/pilot_server.go:109 +0x84
  istio.io/istio/pilot/pkg/proxy/envoy/v2_test.initLocalPilotTestEnv()
      /go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/xds_test.go:140 +0x101
  istio.io/istio/pilot/pkg/proxy/envoy/v2_test.TestAdsReconnectWithNonce()
      /go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/ads_test.go:29 +0x53
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 292 (running) created at:
  istio.io/istio/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:678 +0xb8
  istio.io/istio/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /go/src/istio.io/istio/vendor/google.golang.org/grpc/transport/http2_server.go:416 +0x1443
  istio.io/istio/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /go/src/istio.io/istio/vendor/google.golang.org/grpc/transport/http2_server.go:456 +0x87a
  istio.io/istio/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:676 +0x16e
  istio.io/istio/vendor/google.golang.org/grpc.(*Server).handleRawConn.func2()
      /go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:631 +0x53
  istio.io/istio/vendor/google.golang.org/grpc.(*Server).handleRawConn.func3()
      /go/src/istio.io/istio/vendor/google.golang.org/grpc/server.go:639 +0x34
==================
testing.go:730: race detected during execution of test
```

### PR DESCRIPTION
**Pull request by caohe**
Fixes #11621

### CODE REVIEW

**Comment by caohe** [comment]
Lint fails on unrelated files

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 1]
```diff
@@ -231,10 +231,12 @@ func (s *DiscoveryServer) periodicRefreshMetrics() {
 	defer ticker.Stop()
 	for range ticker.C {
 		push := s.globalPushContext()
+		push.Mutex.Lock()
```
Comment (nmittler) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: It looks like more than fixing a flake here. We're fixing a bug. Do we know when this bug was introduced and why it wasn't caught by our tests?

cc @costinm @rshriram

Comment (caohe) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: Hi @nmittler ,

Could this pr be approved to fix the bug in release-1.0? Or could I do something further?

Thanks.

**Comment by ymesika** [comment]
The racetest is still failing.
Please wait for https://github.com/istio/istio/pull/11616 to be merged, rebase and make sure it passes.

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 1 above)
Comment (ymesika) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: This race fix has been fixed in master/1.1 as part of #10036

**Comment by ymesika** [comment]
Still pending..

**Comment by caohe** [comment]
@ymesika @nmittler  Need this bug be fixed in release-1.0? Could you tell me what the policy is?

**Comment by nmittler** [comment]
cc @duderino for policy on backporting

**Comment by caohe** [comment]
> The racetest is still failing.
> Please wait for #11616 to be merged, rebase and make sure it passes.

@ymesika @geeknoid @nmittler PTAL
#11616 has been merged. I have rebased it. And the racetest passes.

**Comment by caohe** [comment]
@ymesika @geeknoid @nmittler 
I found the racetest failed frequently in the past few days due to this bug. Such as [#368859](https://gubernator.k8s.io/build/istio-circleci/racetest/371783) and [#371552](https://gubernator.k8s.io/build/istio-circleci/racetest/371552).
Could one of you look/approve to fix this bug?

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 1 above)
Comment (nmittler) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: @duderino what is the policy for commit to the release-1.0 branch? Are we still releasing 1.0 patches?

Comment (caohe) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: @nmittler I found #11616 was merged to release-1.0 branch last week.

**Comment by caohe** [comment]
@nmittler Thanks for approval. Could you please merge it?

**Comment by nmittler** [comment]
@duderino are you the gatekeeper for the 1.0 branch?

**Comment by caohe** [comment]
@geeknoid Could you please help to merge this pr? It has been approved.

**Pull request closed without merging**